cmd/crank/beta/trace/internal/resource/xrm/loader_test.go

```diff
@@ -0,0 +1,173 @@
+package xrm
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"regexp"
+	"strconv"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/crossplane/crossplane/cmd/crank/beta/trace/internal/resource"
+)
+
+var reNum = regexp.MustCompile(`-(\d+)$`)
+
+type simpleGenerator struct {
+	childDepth int
+	numItems   int
+}
+
+func (d *simpleGenerator) createResource(apiVersion, kind, name string) *resource.Resource {
+	obj := map[string]any{
+		"apiVersion": apiVersion,
+		"kind":       kind,
+		"metadata": map[string]any{
+			"name": name,
+		},
+	}
+	return &resource.Resource{Unstructured: unstructured.Unstructured{Object: obj}}
+}
+
+func (d *simpleGenerator) createRefAtDepth(depth int) v1.ObjectReference {
+	prefix := "comp-res"
+	if depth == d.childDepth {
+		prefix = "managed-res"
+	}
+	return v1.ObjectReference{
+		Kind:       fmt.Sprintf("Depth%d", depth),
+		Name:       fmt.Sprintf("%s-%d-%d", prefix, rand.Int(), depth),
+		APIVersion: "example.com/v1",
+	}
+}
+
+func (d *simpleGenerator) createResourceFromRef(ref *v1.ObjectReference) *resource.Resource {
+	return d.createResource(ref.APIVersion, ref.Kind, ref.Name)
+}
+
+func (d *simpleGenerator) loadResource(_ context.Context, ref *v1.ObjectReference) *resource.Resource {
+	return d.createResourceFromRef(ref)
+}
+
+func (d *simpleGenerator) depthFromResource(res *resource.Resource) int {
+	ret := 0
+	matches := reNum.FindStringSubmatch(res.Unstructured.GetName())
```

Comment:
remembered later on that we do need something that fulfills the `resourceLoader` interface to use in these tests, so yes an object with some mocked logic is needed instead of just pure data in table driven test cases. but if there's a way to simplify the internal model its using, that may still be nice 😇

Reply:
I agree. This is ugly. Resolved in df75bcaaae93e7a956d44f0c62fe3a7982b70f19 by maintaining an explicit map of resource names to depth.